presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,289 @@
+                (QueryBody) process(node.getQueryBody(), context),
+                node.getOrderBy(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitTableSubquery(TableSubquery node, Void context)
+    {
+        return new TableSubquery((Query) process(node.getQuery(), context));
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        // Process from and select clause before all others to check if query is rewritable and acquire alias relation if any
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
+        Select rewrittenSelect = (Select) process(node.getSelect(), context);
+        Optional<Expression> rewrittenWhere = node.getWhere().map(where -> (Expression) process(where, context));
+        Optional<GroupBy> rewrittenGroupBy = node.getGroupBy().map(groupBy -> (GroupBy) process(groupBy, context));
+        Optional<Expression> rewrittenHaving = node.getHaving().map(having -> (Expression) process(having, context));
+        Optional<OrderBy> rewrittenOrderBy = node.getOrderBy().map(orderBy -> (OrderBy) process(orderBy, context));
+
+        return new QuerySpecification(
+                rewrittenSelect,
+                rewrittenFrom,
+                rewrittenWhere,
+                rewrittenGroupBy,
+                rewrittenHaving,
+                rewrittenOrderBy,
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitSelect(Select node, Void context)
+    {
+        ImmutableList.Builder<SelectItem> rewrittenSelectItems = ImmutableList.builder();
+
+        for (SelectItem selectItem : node.getSelectItems()) {
+            SelectItem rewrittenSelectItem = (SelectItem) process(selectItem, context);
+            rewrittenSelectItems.add(rewrittenSelectItem);
+        }
+
+        return new Select(node.isDistinct(), rewrittenSelectItems.build());
+    }
+
+    @Override
+    protected Node visitSingleColumn(SingleColumn node, Void context)
+    {
+        return new SingleColumn((Expression) process(node.getExpression(), context), node.getAlias());
+    }
+
+    @Override
+    protected Node visitAllColumns(AllColumns node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitArithmeticBinary(ArithmeticBinaryExpression node, Void context)
+    {
+        return new ArithmeticBinaryExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitIdentifier(Identifier node, Void context)
+    {
+        return new Identifier(getViewColumnName(node.getValue()), node.isDelimited());
+    }
+
+    @Override
+    protected Node visitFunctionCall(FunctionCall node, Void context)
+    {
+        String functionCall = node.toString();
```

Comment:
Here, we should be able to find if the expression is communicative and associative,  then we should be able to add some information, to make sure we can look for all variant when looking up keys.
however this logic can be extended in further PRs.